Resources/Public/Javascript/OpenLayers/ImageProcessingUtility.js

```diff
@@ -0,0 +1,225 @@
+ * LICENSE.txt file that was distributed with this source code.
+ */
+
+var imageProcessingUtility = {
+
+  'filter': function(canvas, filters) {
+    if (canvas !== undefined) {
+      var ctx = canvas.getContext('2d');
+      for (var filter in filters) {
+        this.applyFilter(canvas, filter, filters[filter]);
+      }
+    }
+  },
+
+  'applyFilter': function(canvas, filter, value) {
+    var ctx = canvas.getContext('2d');
+    var imgData = ctx.getImageData(0, 0, ctx.canvas.width, ctx.canvas.height);
+    switch (filter) {
+      case 'contrast':
+        imgData = this.contrast(imgData, value);
+        ctx.putImageData(imgData, 0, 0);
+        break;
+      case 'hue':
+        imgData = this.hue(imgData, value);
+        ctx.putImageData(imgData, 0, 0);
+        break;
+      case 'saturation':
+        imgData = this.saturation(imgData, value);
+        ctx.putImageData(imgData, 0, 0);
+        break;
+      case 'brightness':
+        imgData = this.brightness(imgData, value);
+        ctx.putImageData(imgData, 0, 0);
+        break;
+      case 'invert':
+        imgData = this.invert(imgData);
+        ctx.putImageData(imgData, 0, 0);
+        break;
+
+      default:
+        break;
+    }
+  },
+
+  'contrast': function (imgData, value) {
+    var data = imgData.data;
+    var contrast = (value * 100 - 100) * 2.55;
+    var factor = (255 + contrast) / (255.01 - contrast);
+
+    for(var i=0;i<data.length;i+=4)
+    {
+      data[i] = factor * (data[i] - 128) + 128;     //r value
+      data[i+1] = factor * (data[i+1] - 128) + 128; //g value
+      data[i+2] = factor * (data[i+2] - 128) + 128; //b value
+
+    }
+    return imgData;  //optional (e.g. for filter function chaining)
+  },
+
+  'saturation': function (imgData, value) {
+    var data = imgData.data;
+    var sv = value;
+    var luR = 0.3086; // constant to determine luminance of red. Similarly, for green and blue
+    var luG = 0.6094;
+    var luB = 0.0820;
+
+    var az = (1 - sv)*luR + sv;
+    var bz = (1 - sv)*luG;
+    var cz = (1 - sv)*luB;
+    var dz = (1 - sv)*luR;
+    var ez = (1 - sv)*luG + sv;
+    var fz = (1 - sv)*luB;
+    var gz = (1 - sv)*luR;
+    var hz = (1 - sv)*luG;
+    var iz = (1 - sv)*luB + sv;
+
+    for(var i = 0; i < data.length; i += 4)
+    {
+      var red = data[i]; // Extract original red color [0 to 255]. Similarly for green and blue below
+      var green = data[i + 1];
+      var blue = data[i + 2];
+
+      var saturatedRed = (az*red + bz*green + cz*blue);
+      var saturatedGreen = (dz*red + ez*green + fz*blue);
+      var saturateddBlue = (gz*red + hz*green + iz*blue);
+
+      data[i] = saturatedRed;
+      data[i + 1] = saturatedGreen;
+      data[i + 2] = saturateddBlue;
+    }
+
+    return imgData;
+  },
+
+  'brightness': function (imgData, value) {
+    var data = imgData.data;
+    var brightnessMul = value; // brightness multiplier
+
+    for(var i = 0; i < data.length; i += 4)
```

Comment:
Maybe explanation for +4 should be written in some general comment before all those functions?